content/blog/2022-04-21-debug-your-lamda-functions-from-your-ide-using-localstack/index.md

```diff
@@ -0,0 +1,242 @@
+    import time, threading
+    import sys, glob
+    sys.path.append(glob.glob(".venv/lib/python*/site-packages")[0])
+    import debugpy
+
+    debugpy.listen(("0.0.0.0", 19891))
+    class T(threading.Thread):
+        daemon = True
+        def run(self):
+            time.sleep(timeout)
+            print("Canceling debug wait task ...")
+            debugpy.wait_for_client.cancel()
+    T().start()
+    print("Waiting for client to attach debugger ...")
+    debugpy.wait_for_client()
+
+
+if __name__ == "__main__":
+    handler({}, {})
+```
+
+Before debugging it, we first need to create it on LocalStack. We will deploy the Lambda using a unique S3 bucket indicated by `__local__` as the bucket name.
+
+The S3 key path should point to the directory of your Lambda function code. You can save the above example as a file in a directory of your choice. The handler is referenced by the filename of your Lambda function, where the code inside of it is invoked. Push the following command to your terminal:
+
+```sh
+awslocal lambda create-function --function-name my-cool-local-function \
+    --code S3Bucket="__local__",S3Key="$(pwd)/" \
+    --handler handler.handler \
+    --runtime python3.8 \
+    --role cool-stacklifter
+```
+
+We can test out the above Lambda function by specifying a payload:
+
+```sh
+awslocal lambda invoke --function-name my-cool-local-function --payload '{"message": "Hello from LocalStack!"}' output.txt
+```
+Configuring Visual Studio Code for Python Lambda debugging
+
+For attaching the debug server from Visual Studio Code, you need to add a run configuration.
+
+```json
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "Python: Remote Attach",
+            "type": "python",
+            "request": "attach",
+            "connect": {
+                "host": "localhost",
+                "port": 19891
+            },
+            "pathMappings": [
+                {
+                    "localRoot": "${workspaceFolder}",
+                    "remoteRoot": "."
+                }
+            ]
+        }
+    ]
+}
+```
+
+You have about 15 seconds to switch to Visual Studio Code and run the preconfigured remote debugger with the above function. This timeout is configurable, and now you can set a breakpoint in the Lambda handler code first, which can then later be inspected.
+
+Here is our Lambda debugger in action:
+
+{{< img src="lambda-function-debugging-vs-code.png" >}}
+
+## Configuring LocalStack for JVM debugging
+
+To configure LocalStack for JVM debugging, you would need to set the `LAMBDA_JAVA_OPTS` with jdwp settings and expose the debug port of your choice. Over a `docker-compose.yml`, you can do that via:
+
+```yml
+services:
+  localstack:
+    ...
+    environment:
+      ...
+      - LAMBDA_JAVA_OPTS=-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=*:5050
+      - LAMBDA_DOCKER_FLAGS=-p 127.0.0.1:5050:5050
+```
```

Comment:
Above we're using the localstack CLI to start LocalStack, why switch to docker-compose now?

Reply:
The instructions for Python and Java have been different. I have made some changes to reflect the difference in them.